CommunityProject/Assets/_Game/Scripts/Runtime/Entities/Weapons/Weapon.cs

```diff
@@ -11,21 +11,28 @@
 namespace BoundfoxStudios.CommunityProject.Entities.Weapons
 {
 	[SelectionBase]
-	public abstract class Weapon<T> : MonoBehaviour
-		where T : WeaponSO
+	public abstract class Weapon<TWeaponSO, TTargetLocatorSO, TEffectiveWeaponDefinition> : MonoBehaviour
+		where TWeaponSO : WeaponSO
+		where TTargetLocatorSO : TargetLocatorSO<TEffectiveWeaponDefinition>
+		where TEffectiveWeaponDefinition : EffectiveWeaponDefinition
 	{
 		[field: SerializeField]
-		public T WeaponDefinition { get; private set; } = default!;
+		public TWeaponSO WeaponDefinition { get; private set; } = default!;
 
 		[SerializeField]
-		protected TargetLocator<T> TargetLocator = default!;
+		protected TTargetLocatorSO TargetLocator = default!;
+
+		[SerializeField]
+		protected EffectiveWeaponCalculatorSO<TWeaponSO, TEffectiveWeaponDefinition> EffectiveWeaponCalculatorSO = default!;
 
 		[field: SerializeField]
 		public Tower Tower { get; private set; } = default!;
 
 		[field: SerializeField]
 		public TargetType TargetType { get; private set; }
 
+		public TEffectiveWeaponDefinition? EffectiveWeaponDefinition { get; private set; }
```

Comment:
```suggestion
    private TEffectiveWeaponDefinition? _effectiveWeaponDefinition;
    
		public TEffectiveWeaponDefinition EffectiveWeaponDefinition => _effectiveWeaponDefinition ??= CalculateEffectiveWeaponDefinition();
```

Hat den Vorteil, dass der Zugriff auf `EffectiveWeaponDefinition` nicht mehr null zurück geben kann, d.h. die Änderungen am `TargetLocatorSO` mit dem Hinzufügen von `?` könnte man rückgängig machen.

Plus, das müsste man mal ausprobieren, könnten die Gizmos nun auch auf `EffectiveWeaponDefinition` zugreifen anstatt auf `WeaponDefinition`, damit müsste man dann auch im Editor die echte Reichweite sehen.